crates/dekoder/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+//! eko output interface.
+use float_cmp::approx_eq;
+use ndarray::Array4;
+use std::collections::HashMap;
+use std::fs::remove_dir_all;
+use std::fs::File;
+use std::io::BufWriter;
+use std::path::PathBuf;
+use thiserror::Error;
+use yaml_rust2::Yaml;
+
+mod inventory;
+
+/// The EKO errors.
+#[derive(Error, Debug)]
+pub enum EKOError {
+    #[error("No working directory")]
+    NoWorkingDir,
+    #[error("I/O error")]
+    IOError(#[from] std::io::Error),
+    #[error("Loading operator from `{0}` failed")]
+    OperatorLoadError(PathBuf),
+    #[error("Failed to read key(s) `{0}`")]
+    KeyError(String),
+}
+
+/// My result type has always my errros.
+type Result<T> = std::result::Result<T, EKOError>;
+
+/// Default value for the float comparison inside `EvolutionPoint`.
+const EP_CMP_ULPS: i64 = 64;
+
+/// A reference point in the evolution atlas.
+pub struct EvolutionPoint {
+    /// Evolution scale.
+    pub scale: f64,
+    /// Number of flavors
+    pub nf: i64,
+}
+
+/// 4D evolution operator.
+pub struct Operator {
+    pub op: Option<Array4<f64>>,
+}
+
+impl Operator {
+    /// Empty initializer.
+    pub fn new() -> Self {
+        Self { op: None }
+    }
+}
+
+impl Default for Operator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TryFrom<&Yaml> for EvolutionPoint {
+    type Error = EKOError;
+
+    /// Load from yaml.
+    fn try_from(yml: &Yaml) -> Result<Self> {
+        // work around float representation
+        let scale = yml["scale"].as_f64();
+        let scale = if scale.is_some() {
+            scale.ok_or(EKOError::KeyError(
+                "because failed to read scale as float".to_owned(),
+            ))?
+        } else {
+            yml["scale"].as_i64().ok_or(EKOError::KeyError(
+                "because failed to read scale as float from int".to_owned(),
+            ))? as f64
+        };
+        let nf = yml["nf"]
+            .as_i64()
+            .ok_or(EKOError::KeyError("because failed to read nf".to_owned()))?;
+        Ok(Self { scale, nf })
+    }
+}
+
+impl PartialEq for EvolutionPoint {
+    /// (Protected) comparator.
+    fn eq(&self, other: &Self) -> bool {
+        self.nf == other.nf && approx_eq!(f64, self.scale, other.scale, ulps = EP_CMP_ULPS)
+    }
+}
+
+impl Eq for EvolutionPoint {}
+
+/// EKO output
+pub struct EKO {
+    /// Working directory
+    path: PathBuf,
+    /// final operators
+    operators: inventory::Inventory<EvolutionPoint>,
+}
+
+/// Operators directory.
+const DIR_OPERATORS: &str = "operators/";
+
+impl EKO {
+    /// Check our working directory is safe.
+    fn check(&self) -> Result<()> {
+        let path_exists = self.path.try_exists().is_ok_and(|x| x);
+        if !path_exists {
+            return Err(EKOError::NoWorkingDir);
+        }
+        Ok(())
+    }
+
+    /// Remove the working directory.
+    pub fn destroy(&self) -> Result<()> {
+        self.check()?;
+        Ok(remove_dir_all(&self.path)?)
```

Comment:
```suggestion
        remove_dir_all(&self.path)
```

Reply:
Not possible:
```
error[E0308]: mismatched types
   --> crates/dekoder/src/lib.rs:108:9
    |
106 |     pub fn destroy(&self) -> Result<()> {
    |                              ---------- expected `std::result::Result<(), EKOError>` because of return type
107 |         self.assert_working_dir()?;
108 |         remove_dir_all(&self.path)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<(), EKOError>`, found `Result<(), Error>`
    |
    = note: expected enum `std::result::Result<_, EKOError>`
               found enum `std::result::Result<_, std::io::Error>`
help: use `?` to coerce and return an appropriate `Err`, and wrap the resulting value in `Ok` so the expression remains of type `Result`
    |
108 |         Ok(remove_dir_all(&self.path)?)
    |         +++                          ++
```

Reply:
```suggestion
        remove_dir_all(&self.path).into()
```
?

Reply:
Sorry, I just saw the other one https://github.com/NNPDF/eko/pull/400#discussion_r1724625978